src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,105 @@
+package duke.parser;
+import java.util.Scanner;
+
+import duke.exception.DukeException;
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+/**
+ * Parser class
+ * To process the data in String format
+ */
+public class Parser {
+    private Scanner userIn;
+
+    /**
+     * Constructor method for Parser.
+     *
+     * @param userIn Scanner object that takes in user input.
+     */
+    public Parser(Scanner userIn) {
+        this.userIn = userIn;
+    }
+
+    /**
+     * Main method for parsing commands.
+     *
+     * @param taskList TaskList containing list of all tasks in the program.
+     * @param ui ui to show user messages.
+     * @throws DukeException if there is an error (usually user input error).
+     */
```

Comment:
Wow, I really like how this function is organised! Great job being careful with nested code!